sdv/multi_table/hma.py

```diff
@@ -44,6 +44,88 @@ def __init__(self, metadata, locales=None, verbose=True):
             self._table_synthesizers,
             self._table_sizes)
 
+    def _num_extended_columns(self, table_name, parent_name, columns_per_table):
+        """Get the number of columns that will be generated for table_name.
+
+        A table generates:
+            - 1 num_rows column for each for each foreign key with a specific parent
+            - n*(n-1)/2 correlation columns for each data column
+            - 4 parameters columns for each data column, with:
+                - 1 column for parameter a
+                - 1 column for parameter b
+                - 1 column for parameter scale
+                - 1 column for parameter loc
+        """
+        # The `num_rows` column. Because HMA models at most one relationship between
+        # two tables, this is always 1
+        num_cardinality_columns = 1
+
+        # no parameter columns are generated if there are no data columns
+        num_data_columns = columns_per_table[table_name]
+        if num_data_columns == 0:
+            return num_cardinality_columns
+
+        num_correlation_columns = (num_data_columns - 1) * num_data_columns // 2
+        num_parameters_columns = num_data_columns * 4
+
+        return num_correlation_columns + num_cardinality_columns + num_parameters_columns
+
+    def _estimate_columns_traversal(self, table_name, columns_per_table, visited):
+        """Given a table, estimate how many columns each parent will model.
+
+        This method recursively models the children of a table all the way to the leaf nodes.
+        """
+        for child_name in self.metadata._get_child_map()[table_name]:
+            if child_name not in visited:
+                self._estimate_columns_traversal(child_name, columns_per_table, visited)
+
+            columns_per_table[table_name] += \
+                self._num_extended_columns(child_name, table_name, columns_per_table)
+
+        visited.add(table_name)
+
+    def _get_num_data_columns(self):
+        """Get the number of data columns, ie colums that are not id, for each table."""
+        columns_per_table = {}
+        for table_name, table in self.metadata.tables.items():
+            columns_per_table[table_name] = \
+                sum([1 for col in table.columns.values() if col['sdtype'] != 'id'])
+
+        return columns_per_table
+
+    def estimate_number_of_root_columns(self):
```

Comment:
I think we should make this private for now and expose later if we need to. We can also just call it `_estimate_num_columns`